x-pack/plugins/lens/public/indexpattern_datasource/operations/layer_helpers.ts

```diff
@@ -317,34 +354,282 @@ export function replaceColumn({
+      hasExactMatch = true;
+
+      const newLayer = { ...layer, columns: { ...layer.columns, [newId]: { ...previousColumn } } };
+      layer = {
+        ...layer,
+        columnOrder: getColumnOrder(newLayer),
+        columns: adjustColumnReferencesForChangedColumn(newLayer, newId),
+      };
+      return newId;
+    }
+
+    // Look for any fieldless operations that can be inserted directly (case n2)
+    if (validation.input.includes('none')) {
+      const validOperations = operationDefinitions.filter((def) => {
+        if (def.input !== 'none') return;
+        return isOperationAllowedAsReference({
+          validation,
+          operationType: def.type,
+          indexPattern,
+        });
+      });
+
+      if (validOperations.length === 1) {
+        layer = insertNewColumn({
+          layer,
+          columnId: newId,
+          op: validOperations[0].type,
+          indexPattern,
+        });
+        return newId;
+      }
+    }
+
+    // Try to reuse the previous field by finding a possible operation. Because we've alredy
+    // checked for an exact operation match, this is guaranteed to be different from previousColumn
+    if (!hasFieldMatch && 'sourceField' in previousColumn && validation.input.includes('field')) {
+      const defIgnoringfield = operationDefinitions
+        .filter(
+          (def) =>
+            def.input === 'field' &&
+            isOperationAllowedAsReference({ validation, operationType: def.type, indexPattern })
+        )
+        .sort(getSortScoreByPriority);
+
+      // No exact match found, so let's determine that the current field can be reused
+      const defWithField = defIgnoringfield.filter((def) => {
+        if (def.input !== 'field') return;
+        const previousField = indexPattern.getFieldByName(previousColumn.sourceField);
+        if (!previousField) return;
+        return isOperationAllowedAsReference({
+          validation,
+          operationType: def.type,
+          field: previousField,
+          indexPattern,
+        });
+      });
+
+      if (defWithField.length > 0) {
+        // Found the best match that keeps the field (case n3)
+        hasFieldMatch = true;
+        layer = insertNewColumn({
+          layer,
+          columnId: newId,
+          op: defWithField[0].type,
+          indexPattern,
+          field: indexPattern.getFieldByName(previousColumn.sourceField),
+        });
+        return newId;
+      } else if (defIgnoringfield.length === 1) {
+        // Can't use the field, but there is an exact match on the operation (case n4)
+        hasFieldMatch = true;
+        layer = {
+          ...layer,
+          incompleteColumns: {
+            ...layer.incompleteColumns,
+            [newId]: { operationType: defIgnoringfield[0].type },
+          },
+        };
+        return newId;
+      }
+    }
+
+    // Look for field-based references that we can use to assign a new field-based operation from (case r2)
```

Comment:
Case r2 (and the case n5 referenced in line 591) are not explained, did the names change?